CHANGELOG.md

```diff
@@ -5,13 +5,15 @@ UPGRADE NOTES:
 * The official Linux packages for the v1.2 series now require Linux kernel version 2.6.32 or later.
 * When making outgoing HTTPS or other TLS connections as a client, Terraform now requires the server to support TLS v1.2. TLS v1.0 and v1.1 are no longer supported. Any safely up-to-date server should support TLS 1.2, and mainstream web browsers have required it since 2020.
 * When making outgoing HTTPS or other TLS connections as a client, Terraform will no longer accept CA certificates signed using the SHA-1 hash function. Publicly trusted Certificate Authorities have not issued SHA-1 certificates since 2015.
+* The [terraform-credentials-env credentials helper](https://github.com/apparentlymart/terraform-credentials-env) has been implemented as a Terraform feature. Users of terraform-credentials-env should note that the TF_TOKEN_* variables now take priority over CLI config credentials blocks created by `terraform login` for like host names.
```

Comment:
Do we want to link to an external tool here as part of the CHANGELOG? We don't always need to add an entry under "notes", especially if there is already a changelog entry for the feature.

Reply:
This is a kinda-unprecedented situation where we have an upgrade note specifically referring to a "third-party tool" (even though I work on Terraform, I wrote this of my own accord without consulting with the rest of the Terraform team) because we've effectively incorporated that tool into the product itself.

Because of that, I do feel a little torn here: the users of this third-party helper are only a very tiny portion of the total set of Terraform users, but it does still seem notable that they can stop using that credentials helper but that the behavior will be slightly different.

My recommendation would be to keep this in but to rework it a bit so that we're a bit clearer that this point only applies to someone who has been using that credentials helper, thus allowing other readers to quickly notice that this doesn't apply to them. For example:

> * If you use [the third-party credentials helper plugin `terraform-credentials-env`](https://github.com/apparentlymart/terraform-credentials-env), you should disable it as part of upgrading to Terraform v1.2 because similar functionality is now built in to Terraform itself.
>
>     The new behavior supports the same environment variable naming scheme but has a difference in priority order from the credentials helper: `TF_TOKEN_...` environment variables will now take priority over `credentials` blocks in CLI configuration and credentials stored automatically by `terraform login`, which is not true for credentials provided by any credentials helper plugin. If you see Terraform using different credentials after upgrading, check to make sure you do not specify credentials for the same host in multiple locations.
>
>    If you use the credentials helper in conjunction with the `hashicorp/tfe` Terraform provider to manage Terraform Cloud or Terraform Enterprise objects with Terraform, you should also upgrade to version <TBD> of that provider, which added the corresponding built-in support for these environment variables.

My goals with this writing were:
* Since it's an "upgrade note" I tried to make it clearly actionable, leading with: If _X_, do _Y_ as part of upgrading.
* Follow up with some more notes to help people understand some possible but less likely consequences of taking that action, and what to do about them if they occur.
* Also make the connection with the `hashicorp/tfe` provider to try to head off questions like "after I removed the credentials helper, the provider stopped working".

It's more wordy than the original proposal, but we do typically spend more words on the "upgrade notes" than we do in the other changelog sections because there are typically far fewer of them and I think users typically appreciate the extra detail.

As we get closer to release and prepare the "upgrade guide" (which is evolving into more like "curated release notes" now we're post v1.0) I expect we'll include a section in there which will have more room for showing some real examples of what this entails in case someone needs some more information on this. We typically prepare those as a separate step shortly before release, using the changelog as the prompt, so no need to worry too much about that for now.